src/manage_sql_tls_certificates.c

```diff
@@ -447,6 +447,256 @@ copy_tls_certificate (const char *name,
+           " AND resource_location = %i"
+           " AND resource = %llu;",
+           LOCATION_TABLE,
+           tls_certificate);
+
+      tags_remove_resource ("tls_certificate",
+                            tls_certificate,
+                            LOCATION_TABLE);
+    }
+
+  sql ("DELETE FROM tls_certificates WHERE id = %llu;",
+       tls_certificate);
+
+  sql_commit ();
+  return 0;
+}
+
+/**
+ * @brief Try restore a tls_certificate.
+ *
+ * If success, ends transaction for caller before exiting.
+ *
+ * @param[in]  tls_certificate_id  UUID of resource.
+ *
+ * @return 0 success, 1 fail because tls_certificate is in use,
+ *         2 failed to find tls_certificate, -1 error.
+ */
+int
+restore_tls_certificate (const char *tls_certificate_id)
+{
+  tls_certificate_t trash_tls_certificate, tls_certificate;
+
+  if (find_trash ("tls_certificate",
+                  tls_certificate_id,
+                  &trash_tls_certificate))
+    {
+      sql_rollback ();
+      return -1;
+    }
+
+  if (trash_tls_certificate == 0)
+    return 2;
+
+  /* Move the tls_certificate back to the regular table. */
+
+  sql ("INSERT INTO tls_certificates"
+       " (uuid, owner, name, comment, creation_time, modification_time,"
+       "  certificate, subject_dn, issuer_dn, trust,"
+       "  activation_time, expiration_time, md5_fingerprint)"
+       " SELECT"
+       "  uuid, owner, name, comment, creation_time, modification_time,"
+       "  certificate, subject_dn, issuer_dn, trust,"
+       "  activation_time, expiration_time, md5_fingerprint"
+       " FROM tls_certificates_trash WHERE id = %llu;",
+       trash_tls_certificate);
+
+  tls_certificate = sql_last_insert_id ();
+
+  /* Adjust references to the tls_certificate. */
+
+  permissions_set_locations ("tls_certificate",
+                             trash_tls_certificate,
+                             tls_certificate,
+                             LOCATION_TABLE);
+  tags_set_locations ("tls_certificate",
+                      trash_tls_certificate,
+                      tls_certificate,
+                      LOCATION_TABLE);
+
+  /* Clear out the trashcan tls_certificate. */
+
+  sql ("DELETE FROM tls_certificates_trash WHERE id = %llu;",
+       trash_tls_certificate);
+
+  sql_commit ();
+  return 0;
+}
+
+/**
+ * @brief Empty TLS certificate trashcans.
+ */
+void
+empty_trashcan_tls_certificates ()
+{
+  sql ("DELETE FROM permissions"
+       " WHERE resource_type = 'tls_certificate'"
+       " AND resource_location = %i"
+       " AND resource IN (SELECT id FROM tls_certificates_trash"
+       "                  WHERE owner = (SELECT id FROM users"
+       "                                 WHERE uuid = '%s'));",
+       LOCATION_TRASH,
+       current_credentials.uuid);
+
+  sql ("DELETE FROM tickets_trash"
+       " WHERE owner = (SELECT id FROM users WHERE uuid = '%s');",
+       current_credentials.uuid);
+}
+
+/**
+ * @brief Delete all TLS certificate owner by a user.
```

Comment:
"owned"

Reply:
Fixed in 106673d51b67f0c62e005f3a64236f27b81d4586